File: OnlineQuizJava/app/src/main/java/com/example/onlinequizjava/fragments/OwnerRoomFragment.java
package com.example.onlinequizjava.fragments;

import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.onlinequizjava.MainActivity;
import com.example.onlinequizjava.R;
import com.example.onlinequizjava.network.SocketEventListener;
import com.example.onlinequizjava.network.SocketMessageFactory;
import com.example.onlinequizjava.network.WebSocketManager;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class OwnerRoomFragment extends Fragment implements SocketEventListener {

    /*
        Oda sahibi ekranı.

        Bu ekranda:
        - Oda kodu gösterilir
        - Oyuncu listesi gösterilir
        - Dinamik şık eklenebilir
        - Soru server'a gönderilir
        - Quiz başlatılır
    */

    private TextView txtRoomCode;
    private TextView txtInfo;
    private TextView txtPlayers;
    private TextView txtQuestionCount;
    private TextView txtStatus;

    private EditText edtQuestionText;
    private LinearLayout optionsContainer;

    private String roomCode = "";
    private String username = "";
    private int questionTime = 20;

    private int questionCount = 0;

    private final List<OptionRow> optionRows = new ArrayList<>();

    private int selectedOptionIndex = -1;

    public OwnerRoomFragment() {
        super(R.layout.fragment_owner_room);
    }

    public static OwnerRoomFragment newInstance(
            String roomCode,
            String username,
            int questionTime
    ) {
        OwnerRoomFragment fragment = new OwnerRoomFragment();

        Bundle bundle = new Bundle();
        bundle.putString("roomCode", roomCode);
        bundle.putString("username", username);
        bundle.putInt("questionTime", questionTime);

        fragment.setArguments(bundle);

        return fragment;
    }

    private static class OptionRow {
        LinearLayout rowLayout;
        RadioButton radioButton;
        EditText editText;
        Button removeButton;

        OptionRow(
                LinearLayout rowLayout,
                RadioButton radioButton,
                EditText editText,
                Button removeButton
        ) {
            this.rowLayout = rowLayout;
            this.radioButton = radioButton;
            this.editText = editText;
            this.removeButton = removeButton;
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle args = requireArguments();

        roomCode = args.getString("roomCode", "");
        username = args.getString("username", "");
        questionTime = args.getInt("questionTime", 20);
    }

    @Override
    public void onViewCreated(
            @NonNull View view,
            @Nullable Bundle savedInstanceState
    ) {
        super.onViewCreated(view, savedInstanceState);

        txtRoomCode = view.findViewById(R.id.txtRoomCode);
        txtInfo = view.findViewById(R.id.txtInfo);
        txtPlayers = view.findViewById(R.id.txtPlayers);
        txtQuestionCount = view.findViewById(R.id.txtQuestionCount);
        txtStatus = view.findViewById(R.id.txtStatus);

        edtQuestionText = view.findViewById(R.id.edtQuestionText);
        optionsContainer = view.findViewById(R.id.optionsContainer);

        Button btnAddOption = view.findViewById(R.id.btnAddOption);
        Button btnAddQuestion = view.findViewById(R.id.btnAddQuestion);
        Button btnStartQuiz = view.findViewById(R.id.btnStartQuiz);

        WebSocketManager.getInstance().setListener(this);

        txtRoomCode.setText("Oda Kodu: " + roomCode);

        txtInfo.setText(
                "Kullanıcı: " + username + "\n" +
                        "Soru Süresi: " + questionTime + " saniye\n\n" +
                        "Bu kodu diğer kullanıcılara ver.\n" +
                        "Onlar bu kod ile odaya katılacak."
        );

        /*
            Başlangıçta minimum 2 şık açıyoruz.
        */
        addOptionRow("");
        addOptionRow("");

        btnAddOption.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addOptionRow("");
            }
        });

        btnAddQuestion.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addQuestion();
            }
        });

        btnStartQuiz.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startQuiz();
            }
        });
    }

    private void addOptionRow(String defaultText) {
        LinearLayout rowLayout = new LinearLayout(requireContext());
        rowLayout.setOrientation(LinearLayout.HORIZONTAL);
        rowLayout.setGravity(Gravity.CENTER_VERTICAL);
        rowLayout.setPadding(0, dpToPx(8), 0, dpToPx(8));

        LinearLayout.LayoutParams rowParams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        );

        rowLayout.setLayoutParams(rowParams);

        RadioButton radioButton = new RadioButton(requireContext());

        LinearLayout.LayoutParams radioParams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        );

        radioButton.setLayoutParams(radioParams);

        EditText editText = new EditText(requireContext());
        editText.setHint("Şık " + (optionRows.size() + 1));
        editText.setText(defaultText);
        editText.setSingleLine(true);
        editText.setPadding(dpToPx(14), 0, dpToPx(14), 0);
        editText.setBackgroundColor(0xFFFFFFFF);

        LinearLayout.LayoutParams editParams = new LinearLayout.LayoutParams(
                0,
                dpToPx(54),
                1f
        );

        editParams.setMargins(dpToPx(8), 0, dpToPx(8), 0);
        editText.setLayoutParams(editParams);

        Button btnRemove = new Button(requireContext());
        btnRemove.setText("Sil");
        btnRemove.setAllCaps(false);
        btnRemove.setTextSize(13f);

        LinearLayout.LayoutParams removeParams = new LinearLayout.LayoutParams(
                dpToPx(72),
                dpToPx(54)
        );

        btnRemove.setLayoutParams(removeParams);

        rowLayout.addView(radioButton);
        rowLayout.addView(editText);
        rowLayout.addView(btnRemove);

        optionsContainer.addView(rowLayout);

        OptionRow optionRow = new OptionRow(
                rowLayout,
                radioButton,
                editText,
                btnRemove
        );

        optionRows.add(optionRow);

        radioButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                selectOption(optionRow);
            }
        });

        btnRemove.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                removeOptionRow(optionRow);
            }
        });

        updateOptionHints();
    }

    private void selectOption(OptionRow selectedRow) {
        for (int i = 0; i < optionRows.size(); i++) {
            OptionRow row = optionRows.get(i);

            boolean selected = row == selectedRow;
            row.radioButton.setChecked(selected);

            if (selected) {
                selectedOptionIndex = i;
            }
        }
    }

    private void removeOptionRow(OptionRow optionRow) {
        if (optionRows.size() <= 2) {
            txtStatus.setText("En az 2 şık kalmalı.");
            return;
        }

        int removedIndex = optionRows.indexOf(optionRow);

        optionsContainer.removeView(optionRow.rowLayout);
        optionRows.remove(optionRow);

        if (selectedOptionIndex == removedIndex) {
            selectedOptionIndex = -1;
            clearRadioSelection();
        } else if (selectedOptionIndex > removedIndex) {
            selectedOptionIndex--;
        }

        updateOptionHints();
    }

    private void clearRadioSelection() {
        for (OptionRow row : optionRows) {
            row.radioButton.setChecked(false);
        }
    }

    private void updateOptionHints() {
        for (int i = 0; i < optionRows.size(); i++) {
            optionRows.get(i).editText.setHint("Şık " + (i + 1));
        }
    }

    private void addQuestion() {
        String questionText = edtQuestionText.getText().toString().trim();

        if (questionText.isEmpty()) {
            txtStatus.setText("Soru metni boş olamaz.");
            return;
        }

        if (selectedOptionIndex == -1) {
            txtStatus.setText("Doğru cevabı seçmelisin.");
            return;
        }

        List<String> filledOptions = new ArrayList<>();
        int correctIndexInFilledOptions = -1;

        for (int i = 0; i < optionRows.size(); i++) {
            String optionText = optionRows.get(i).editText.getText().toString().trim();

            if (!optionText.isEmpty()) {
                if (i == selectedOptionIndex) {
                    correctIndexInFilledOptions = filledOptions.size();
                }

                filledOptions.add(optionText);
            }
        }

        if (filledOptions.size() < 2) {
            txtStatus.setText("En az 2 dolu şık girmelisin.");
            return;
        }

        if (correctIndexInFilledOptions == -1) {
            txtStatus.setText("Doğru cevap olarak seçtiğin şık boş olamaz.");
            return;
        }

        String message = SocketMessageFactory.addQuestion(
                roomCode,
                questionText,
                filledOptions,
                correctIndexInFilledOptions
        );

        WebSocketManager.getInstance().sendMessage(message);

        txtStatus.setText("Soru gönderildi...");
    }

    private void clearQuestionForm() {
        edtQuestionText.setText("");

        optionsContainer.removeAllViews();
        optionRows.clear();
        selectedOptionIndex = -1;

        addOptionRow("");
        addOptionRow("");
    }

    private void startQuiz() {
        if (questionCount <= 0) {
            txtStatus.setText("Quiz başlatmak için en az 1 soru eklemelisin.");
            return;
        }

        String message = SocketMessageFactory.startQuiz(roomCode);

        WebSocketManager.getInstance().sendMessage(message);

        txtStatus.setText("Quiz başlatma isteği gönderildi...");
    }

    @Override
    public void onSocketConnected() {
        /*
            Bu ekrana gelindiğinde bağlantı genelde zaten açıktır.
        */
    }

    @Override
    public void onSocketMessage(String message) {
        try {
            JSONObject json = new JSONObject(message);
            String type = json.optString("type");

            if ("player_list_updated".equals(type)) {
                JSONArray players = json.optJSONArray("players");
                txtPlayers.setText(buildPlayerText(players));

            } else if ("question_added".equals(type)) {
                questionCount = json.optInt("question_count", questionCount + 1);

                txtQuestionCount.setText("Eklenen soru: " + questionCount);
                txtStatus.setText(json.optString("message", "Soru eklendi."));

                clearQuestionForm();

            } else if ("room_question_count_updated".equals(type)) {
                questionCount = json.optInt("question_count", questionCount);
                txtQuestionCount.setText("Eklenen soru: " + questionCount);

            } else if ("quiz_started".equals(type)) {
                txtStatus.setText("Quiz başladı.");

                ((MainActivity) requireActivity()).openQuizFragment(
                        roomCode,
                        username,
                        questionTime,
                        true
                );

            } else if ("error".equals(type)) {
                txtStatus.setText(json.optString("message", "Bilinmeyen hata oluştu."));
            }

        } catch (Exception e) {
            txtStatus.setText("JSON okuma hatası: " + e.getMessage());
        }
    }

    private String buildPlayerText(JSONArray players) {
        if (players == null || players.length() == 0) {
            return "Oyuncular bekleniyor...";
        }

        StringBuilder builder = new StringBuilder();
        builder.append("Oyuncular:\n\n");

        for (int i = 0; i < players.length(); i++) {
            builder.append(i + 1)
                    .append(". ")
                    .append(players.optString(i))
                    .append("\n");
        }

        return builder.toString();
    }

    @Override
    public void onSocketDisconnected() {
        txtStatus.setText("Sunucu bağlantısı kapandı.");
    }

    @Override
    public void onSocketError(String error) {
        txtStatus.setText("Bağlantı hatası: " + error);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        WebSocketManager.getInstance().removeListener(this);
    }

    private int dpToPx(int dp) {
        return (int) (dp * getResources().getDisplayMetrics().density);
    }
}

File: OnlineQuizKotlin/app/src/main/java/com/example/onlinequizkotlin/fragments/OwnerRoomFragment.kt
package com.example.onlinequizkotlin.fragments

import com.example.onlinequizkotlin.MainActivity
import com.example.onlinequizkotlin.R
import com.example.onlinequizkotlin.network.SocketMessageFactory
import com.example.onlinequizkotlin.network.WebSocketManager
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.example.onlinequizkotlin.network.SocketEventListener
import org.json.JSONArray
import org.json.JSONObject

class OwnerRoomFragment : Fragment(R.layout.fragment_owner_room), SocketEventListener {

    /*
        Oda sahibi ekranı.

        Bu sürümde şıklar artık sabit değildir.

        Önceki yapıda:
        A, B, C, D, E sabitti.

        Bu yapıda:
        - Kullanıcı + Şık Ekle butonuna basar.
        - Her basışta yeni bir şık satırı oluşturulur.
        - Her şık satırında:
            1. RadioButton
            2. EditText
            3. Sil butonu
          bulunur.

        Böylece oda sahibi istediği kadar şıklı soru oluşturabilir.
    */

    private lateinit var txtRoomCode: TextView
    private lateinit var txtInfo: TextView
    private lateinit var txtPlayers: TextView
    private lateinit var txtQuestionCount: TextView
    private lateinit var txtStatus: TextView

    private lateinit var edtQuestionText: EditText
    private lateinit var optionsRadioGroup: RadioGroup

    private var roomCode: String = ""
    private var username: String = ""
    private var questionTime: Int = 20

    private var questionCount: Int = 0

    /*
        Şık satırlarını takip etmek için liste tutuyoruz.

        Her OptionRow:
        - RadioButton
        - EditText
        - Ana satır LinearLayout

        bilgilerini saklar.
    */
    private val optionRows = mutableListOf<OptionRow>()

    companion object {
        fun newInstance(
            roomCode: String,
            username: String,
            questionTime: Int
        ): OwnerRoomFragment {
            val fragment = OwnerRoomFragment()

            val bundle = Bundle()
            bundle.putString("roomCode", roomCode)
            bundle.putString("username", username)
            bundle.putInt("questionTime", questionTime)

            fragment.arguments = bundle

            return fragment
        }
    }

    data class OptionRow(
        val rowLayout: LinearLayout,
        val radioButton: RadioButton,
        val editText: EditText
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        roomCode = requireArguments().getString("roomCode", "")
        username = requireArguments().getString("username", "")
        questionTime = requireArguments().getInt("questionTime", 20)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        txtRoomCode = view.findViewById(R.id.txtRoomCode)
        txtInfo = view.findViewById(R.id.txtInfo)
        txtPlayers = view.findViewById(R.id.txtPlayers)
        txtQuestionCount = view.findViewById(R.id.txtQuestionCount)
        txtStatus = view.findViewById(R.id.txtStatus)

        edtQuestionText = view.findViewById(R.id.edtQuestionText)
        optionsRadioGroup = view.findViewById(R.id.optionsRadioGroup)

        val btnAddOption = view.findViewById<Button>(R.id.btnAddOption)
        val btnAddQuestion = view.findViewById<Button>(R.id.btnAddQuestion)
        val btnStartQuiz = view.findViewById<Button>(R.id.btnStartQuiz)

        WebSocketManager.setListener(this)

        txtRoomCode.text = "Oda Kodu: $roomCode"

        txtInfo.text = """
            Kullanıcı: $username
            Soru Süresi: $questionTime saniye
            
            Bu kodu diğer kullanıcılara ver.
            Onlar bu kod ile odaya katılacak.
        """.trimIndent()

        /*
            Başlangıçta en az 2 şık hazır gelsin.
            Çünkü çoktan seçmeli soruda minimum 2 seçenek gerekir.
        */
        addOptionRow()
        addOptionRow()

        btnAddOption.setOnClickListener {
            addOptionRow()
        }

        btnAddQuestion.setOnClickListener {
            addQuestion()
        }

        btnStartQuiz.setOnClickListener {
            startQuiz()
        }
    }

    private fun addOptionRow(defaultText: String = "") {
        /*
            Şık satırını programatik olarak oluşturuyoruz.

            XML'de sabit EditText koymak yerine her şık burada dinamik üretiliyor.
        */

        val context = requireContext()

        val rowLayout = LinearLayout(context)
        rowLayout.orientation = LinearLayout.HORIZONTAL
        rowLayout.gravity = Gravity.CENTER_VERTICAL
        rowLayout.setPadding(0, 8, 0, 8)

        val rowParams = RadioGroup.LayoutParams(
            RadioGroup.LayoutParams.MATCH_PARENT,
            RadioGroup.LayoutParams.WRAP_CONTENT
        )

        rowLayout.layoutParams = rowParams

        val radioButton = RadioButton(context)

        /*
            Her RadioButton için benzersiz id oluşturuyoruz.

            RadioGroup tek seçim mantığını bu id üzerinden yönetir.
        */
        radioButton.id = View.generateViewId()

        val radioParams = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.WRAP_CONTENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
        )

        radioButton.layoutParams = radioParams

        val editText = EditText(context)
        editText.hint = "Şık ${optionRows.size + 1}"
        editText.setText(defaultText)
        editText.isSingleLine = true
        editText.setPadding(14, 0, 14, 0)
        editText.setBackgroundColor(0xFFFFFFFF.toInt())

        val editParams = LinearLayout.LayoutParams(
            0,
            dpToPx(54),
            1f
        )

        editParams.setMargins(8, 0, 8, 0)

        editText.layoutParams = editParams

        val btnRemove = Button(context)
        btnRemove.text = "Sil"
        btnRemove.isAllCaps = false
        btnRemove.textSize = 13f

        val removeParams = LinearLayout.LayoutParams(
            dpToPx(72),
            dpToPx(54)
        )

        btnRemove.layoutParams = removeParams

        rowLayout.addView(radioButton)
        rowLayout.addView(editText)
        rowLayout.addView(btnRemove)

        optionsRadioGroup.addView(rowLayout)

        val optionRow = OptionRow(
            rowLayout = rowLayout,
            radioButton = radioButton,
            editText = editText
        )

        optionRows.add(optionRow)

        /*
            Sil butonu:
            - Eğer sadece 2 şık kaldıysa silmeye izin vermiyoruz.
            - Çünkü server tarafında en az 2 seçenek istiyoruz.
        */
        btnRemove.setOnClickListener {
            removeOptionRow(optionRow)
        }

        updateOptionHints()
    }

    private fun removeOptionRow(optionRow: OptionRow) {
        if (optionRows.size <= 2) {
            txtStatus.text = "En az 2 şık kalmalı."
            return
        }

        /*
            Silinen şık doğru cevap olarak seçiliyse RadioGroup seçimini temizliyoruz.
        */
        if (optionsRadioGroup.checkedRadioButtonId == optionRow.radioButton.id) {
            optionsRadioGroup.clearCheck()
        }

        optionsRadioGroup.removeView(optionRow.rowLayout)
        optionRows.remove(optionRow)

        updateOptionHints()
    }

    private fun updateOptionHints() {
        /*
            Şık silindikçe veya eklendikçe hint yazıları güncellenir.

            Örnek:
            Şık 1
            Şık 2
            Şık 3
        */
        optionRows.forEachIndexed { index, row ->
            row.editText.hint = "Şık ${index + 1}"
        }
    }

    private fun addQuestion() {
        val questionText = edtQuestionText.text.toString().trim()

        if (questionText.isEmpty()) {
            txtStatus.text = "Soru metni boş olamaz."
            return
        }

        /*
            Boş olmayan şıkları topluyoruz.

            Örneğin 5 şık satırı olabilir ama kullanıcı sadece 3 tanesini doldurmuş olabilir.
            Bu durumda sadece dolu 3 şık server'a gönderilir.
        */
        val filledOptions = mutableListOf<String>()

        /*
            selectedOriginalIndex:
            Kullanıcının RadioButton ile seçtiği şıkkın optionRows içindeki gerçek indexidir.
        */
        val selectedOriginalIndex = getSelectedOptionIndex()

        if (selectedOriginalIndex == -1) {
            txtStatus.text = "Doğru cevabı seçmelisin."
            return
        }

        var correctIndexInFilledOptions = -1

        optionRows.forEachIndexed { originalIndex, row ->
            val optionText = row.editText.text.toString().trim()

            if (optionText.isNotEmpty()) {
                /*
                    Eğer bu satır seçili doğru cevapsa,
                    server'a gönderilecek dolu şıklar listesindeki indexini hesaplıyoruz.
                */
                if (originalIndex == selectedOriginalIndex) {
                    correctIndexInFilledOptions = filledOptions.size
                }

                filledOptions.add(optionText)
            }
        }

        if (filledOptions.size < 2) {
            txtStatus.text = "En az 2 dolu şık girmelisin."
            return
        }

        /*
            Kullanıcı doğru cevap olarak boş bir şıkkı seçmiş olabilir.
            Bu durumda correctIndexInFilledOptions -1 kalır.
        */
        if (correctIndexInFilledOptions == -1) {
            txtStatus.text = "Doğru cevap olarak seçtiğin şık boş olamaz."
            return
        }

        val message = SocketMessageFactory.addQuestion(
            roomCode = roomCode,
            questionText = questionText,
            options = filledOptions,
            correctIndex = correctIndexInFilledOptions
        )

        WebSocketManager.sendMessage(message)

        txtStatus.text = "Soru gönderildi..."
    }

    private fun getSelectedOptionIndex(): Int {
        val checkedId = optionsRadioGroup.checkedRadioButtonId

        if (checkedId == -1) {
            return -1
        }

        optionRows.forEachIndexed { index, row ->
            if (row.radioButton.id == checkedId) {
                return index
            }
        }

        return -1
    }

    private fun clearQuestionForm() {
        edtQuestionText.setText("")

        optionsRadioGroup.removeAllViews()
        optionRows.clear()
        optionsRadioGroup.clearCheck()

        /*
            Yeni soru için tekrar minimum 2 boş şık açıyoruz.
        */
        addOptionRow()
        addOptionRow()
    }

    private fun startQuiz() {
        if (questionCount <= 0) {
            txtStatus.text = "Quiz başlatmak için en az 1 soru eklemelisin."
            return
        }

        val message = SocketMessageFactory.startQuiz(roomCode)
        WebSocketManager.sendMessage(message)

        txtStatus.text = "Quiz başlatma isteği gönderildi..."
    }

    override fun onSocketConnected() {
        // Bu ekrana gelindiğinde bağlantı genelde zaten açıktır.
    }

    override fun onSocketMessage(message: String) {
        val json = JSONObject(message)
        val type = json.optString("type")

        when (type) {
            "player_list_updated" -> {
                val players = json.optJSONArray("players")
                txtPlayers.text = buildPlayerText(players)
            }

            "question_added" -> {
                questionCount = json.optInt("question_count", questionCount + 1)

                txtQuestionCount.text = "Eklenen soru: $questionCount"
                txtStatus.text = json.optString("message", "Soru eklendi.")

                clearQuestionForm()
            }

            "room_question_count_updated" -> {
                questionCount = json.optInt("question_count", questionCount)
                txtQuestionCount.text = "Eklenen soru: $questionCount"
            }

            "quiz_started" -> {
                txtStatus.text = "Quiz başladı."

                (requireActivity() as MainActivity).openQuizFragment(
                    roomCode = roomCode,
                    username = username,
                    questionTime = questionTime,
                    isOwner = true
                )
            }

            "error" -> {
                txtStatus.text = json.optString("message", "Bilinmeyen hata oluştu.")
            }
        }
    }

    private fun buildPlayerText(players: JSONArray?): String {
        if (players == null || players.length() == 0) {
            return "Oyuncular bekleniyor..."
        }

        val builder = StringBuilder()
        builder.append("Oyuncular:\n\n")

        for (i in 0 until players.length()) {
            builder.append("${i + 1}. ${players.optString(i)}\n")
        }

        return builder.toString()
    }

    override fun onSocketDisconnected() {
        txtStatus.text = "Sunucu bağlantısı kapandı."
    }

    override fun onSocketError(error: String) {
        txtStatus.text = "Bağlantı hatası: $error"
    }

    override fun onDestroyView() {
        super.onDestroyView()
        WebSocketManager.removeListener(this)
    }

    private fun dpToPx(dp: Int): Int {
        return (dp * resources.displayMetrics.density).toInt()
    }
}